<a href="https://colab.research.google.com/github/olesia-za/python_for_ds_tasks/blob/main/OZ_Done_hw_3_3_%D0%A0%D0%BE%D0%B7%D1%88%D0%B8%D1%80%D0%B5%D0%BD%D1%96_%D0%BC%D0%B5%D1%82%D0%BE%D0%B4%D0%B8_%D0%BE%D0%B1%D1%80%D0%BE%D0%B1%D0%BA%D0%B8_%D0%B4%D0%B0%D0%BD%D0%B8%D1%85_%D1%83_Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Домашнє завдання до теми "Розширені методи обробки даних у Pandas"

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [2]:
import pandas as pd
df = pd.read_csv('../data/supermarket_sales.csv')

1. Відсортуйте дані за спаданням ціни товара ('Unit price') і виведіть 5 перших рядочків на екран.

In [ ]:
df['Unit price'].nlargest(5) 
# .sort_values(ascending=False)[:5] is taken into account already by nlargest

122    99.96
983    99.96
930    99.92
494    99.89
283    99.83
Name: Unit price, dtype: float64

2. Знайдіть інвойс (рядок в даних) в категорії товарів `Electronic accessories`, в якому був куплений товар з найбільшою кількістю і найбільшою ціною. Виведіть цей рядок даних на екран.

In [4]:
df[df['Product line'] == 'Electronic accessories'].sort_values(by =['Quantity', 'Unit price'], ascending = [False, False]).head(1)

Invoice ID Branch    City Customer type Gender            Product line  \
105  704-48-3927      A  Yangon        Member   Male  Electronic accessories   

     Unit price  Quantity  Tax 5%    Total       Date   Time  Payment   cogs  \
105       88.67        10  44.335  931.035  1/12/2019  14:50  Ewallet  886.7   

     gross margin percentage  gross income  Rating  
105                 4.761905        44.335     7.3

In [5]:
# or alternatively:
df[df['Product line'] == 'Electronic accessories'].nlargest(1, ['Quantity', 'Unit price'])

Invoice ID Branch    City Customer type Gender            Product line  \
105  704-48-3927      A  Yangon        Member   Male  Electronic accessories   

     Unit price  Quantity  Tax 5%    Total       Date   Time  Payment   cogs  \
105       88.67        10  44.335  931.035  1/12/2019  14:50  Ewallet  886.7   

     gross margin percentage  gross income  Rating  
105                 4.761905        44.335     7.3

3.1. Створіть нову колонку в даних `Rating Category`, яка визначає категорію рейтингу (`Rating`): високий, середній чи низький за наступним правилом:
- якщо рейтинг вищий або рівний 8 - то він високий (`High`)
- якщо рейтинг від 6 до 8 не включаючи - то він середній (`Medium`)
- інакше - низький (`Low`).

In [6]:
df['Rating category'] = df.Rating.apply(lambda x: 'High' if x >= 8 else ('Medium' if x >= 6 else 'Low'))
df.head(2)

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   

             Product line  Unit price  Quantity   Tax 5%     Total      Date  \
0       Health and beauty       74.69         7  26.1415  548.9715  1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200  3/8/2019   

    Time  Payment    cogs  gross margin percentage  gross income  Rating  \
0  13:08  Ewallet  522.83                 4.761905       26.1415     9.1   
1  10:29     Cash   76.40                 4.761905        3.8200     9.6   

  Rating category  
0            High  
1            High

3.2. Виведіть, який відсоток даних становить кожне зі значень категорій рейтингу.

In [7]:
df['Rating category'].value_counts()/df['Rating category'].count() * 100

Rating category
Medium    35.6
High      32.9
Low       31.5
Name: count, dtype: float64

4.1. Запустіть код нижче аби створити змінну `time`. Ця змінна містить час в форматі "година:хвилина". Напишіть функцію `get_hour`, яка буде зі змінної формату як `time` дістати лише години і конвертувати їх у тип `int`.

Запустіть написану функцію на змінній `time` та перевірте правильність роботи функції.

In [8]:
time = '11:28'

def get_hour(time):
    hour = int(time.split(':')[0])
    return hour

In [9]:
get_hour(time)

11

4.2. З допомогою функції `get_hour` і метода `DataFrame.apply()` створіть нову колонку `Hour` в нашому наборі даних, яка буде містити години здійснення покупки, обчислені на основі часу здійснення покупки `Time`.

In [10]:
df['Hour'] = df['Time'].apply(get_hour)
df.head(2)

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   

             Product line  Unit price  Quantity   Tax 5%     Total      Date  \
0       Health and beauty       74.69         7  26.1415  548.9715  1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200  3/8/2019   

    Time  Payment    cogs  gross margin percentage  gross income  Rating  \
0  13:08  Ewallet  522.83                 4.761905       26.1415     9.1   
1  10:29     Cash   76.40                 4.761905        3.8200     9.6   

  Rating category  Hour  
0            High    13  
1            High    10

4.3. Виведіть кількість значень за кожною годиною покупки і дайте відповідь "О яких годинах найбільше здійснюють покупок покупці і чому на Ваш погляд саме о цих годинах? Чи це слогічно, що найбільше покупок саме в ці години?".

In [11]:
df.Hour.value_counts().sort_values(ascending=False)
# о 19 найбільше покупок - можна пояснити часом після роботи ( о 18 ще значна кількільсть людей на роботі або не дісьалися до магазину і після 20 вже вдома)
# о 13 - можна пояснити обідом і вільним від роботи часом та можливістю здійснити покупки
# о 15 та 10 дещо дивно, але можливо значан кількість людей в цій місцевості закінчує роботу до 15 або в цей час діє якась година знижок
# щодо 10ої можливо в цей час відкриваються магазини - тому є певний пік покупок або так само діє година знижок


Hour
19    113
13    103
15    102
10    101
18     93
11     90
12     89
14     83
16     77
20     75
17     74
Name: count, dtype: int64

5. Виведіть середнє значення чеку (`Total`) для покупок в містах, наявних в даних, за допомогою функції `groupby`.

In [12]:
df.groupby(['City']).Total.mean()

City
Mandalay     319.872506
Naypyitaw    337.099715
Yangon       312.354031
Name: Total, dtype: float64

6. Виведіть кількість покупок, середнє, мінімальне та максимальне значення для загальної суми чеку (Total) в містах, наявних в даних, за допомогою функції groupby.
Зробіть висновки з цих показників та дайте рекомендації для мережі супермаркетів.

In [14]:
df.groupby(['City']).Total.agg(['size', 'mean', 'min','max', 'sum']).sort_values(by='mean', ascending=False).round(2)
# у Naypyitaw найбільший середній чек та виручка, але найменше покупок - можливо це пов'язано з тим, що там менше населення але більші доходи та/або більше купують дорогих товарів
# у Yangon найбільше покупок, але середній чек менший ніж у Naypyitaw та Mandalay, можливо варто додати акції на дорогі товари або розширити асортимент
#  загалом треба більше інформації по групам товарів, щоб зробити більш точні висновки

size    mean    min      max        sum
City                                              
Naypyitaw   328  337.10  10.68  1042.65  110568.71
Mandalay    332  319.87  18.64  1022.49  106197.67
Yangon      340  312.35  12.69  1039.29  106200.37

7. Ми внесли деякі зміни в наш набір даних, тож, давайте запишемо результат у файл. Запишіть оновлену версію даних у файл `supermarket_sales_enriched.csv`. Зверніть увагу, аби не записати нічого зайвого, наприклад, індекси рядків нам в файлі не потрібні, оскільки вони стандартні.

In [15]:
df.to_csv('../data/supermarket_sales_enriched.csv', index=False)

Найбільше покупок відбувається о 19:00 — 113 разів. Це може бути пов'язано з тим, що багато людей здійснюють покупки після роботи чи навчання, ввечері плануючи вечерю або інші домашні потреби. Також це може бути час, коли супермаркети пропонують спеціальні знижки або акції на кінець дня, щоб збільшити продажі.

В години обідньої перерви, о 13:00, також спостерігається велика кількість покупок — 103. Це може свідчити про те, що люди використовують свою перерву на обід, щоб зробити швидкі покупки або купити щось необхідне в середині робочого дня.

Відносно високі показники покупок о 15:00 — 102 покупки — можуть бути пов'язані з завершенням робочого дня для деяких людей, або з початком знижок та акцій, які часто стартують у другій половині дня.

Що стосується 10:00 ранку з кількістю покупок 101, це може бути пік ранкових покупок, коли люди, можливо, купують продукти на день або виконують планові покупки, оскільки багато супермаркетів починають роботу о цій годині. Це також час, коли супермаркети можуть бути менш забиті, що дозволяє швидше та комфортніше робити покупки.

Висновки:

Naypyitaw має найвищу середню суму чеку (337.10), що може вказувати на більш високу купівельну спроможність або на те, що там продаються дорожчі товари. На це вказує і той факт, що тут найвищий максимальний чек.
Mandalay має середню суму чеку, схожу на Yangon (319.87), але із нижчим максимальним чеком (1022.49), що може вказувати на більш однорідні покупки.
Рекомендації:

Naypyitaw:

Оскільки Naypyitaw має найвищу середню і максимальну суму чеків, маркетингові стратегії можуть бути спрямовані на преміальний сегмент ринку. Пропозиція ексклюзивних товарів або брендів, які приваблюють клієнтів з вищою купівельною спроможністю, може ще більше підвищити продажі.
Запровадження спеціальних подій чи заходів, що пропонують унікальний досвід покупок, може допомогти зберегти існуючих клієнтів і залучити нових.
Розробка програм лояльності, які нагороджують клієнтів за частіші та більші покупки, може заохотити збільшення середнього чеку.
Mandalay:

З огляду на нижчий максимальний чек, супермаркет у Mandalay може зосередитись на пропозиціях, які підкреслюють цінність і доступність, щоб привабити більше клієнтів, які здійснюють регулярні покупки.
Проведення акцій і знижок може стимулювати клієнтів купувати більше за один візит, збільшуючи середню суму чеку.
Вивчення попиту на певні категорії товарів і адаптація асортименту також може допомогти підвищити суму середнього чеку.
Додаткові рекомендації:

Для обох міст: Важливо провести детальний аналіз покупців, щоб зрозуміти їхні переваги та поведінку. Використання CRM(Customer Relationship Management)-систем може допомогти зібрати та аналізувати дані про клієнтів, щоб створювати персоналізовані пропозиції.
Оцінка запасів та управління ними: Ефективне управління запасами може допомогти забезпечити наявність популярних товарів і своєчасне видалення непопулярних, щоб максимізувати прибуток і мінімізувати втрати.
Аналіз конкурентів: Розуміння того, як конкуренти впливають на продажі, може допомогти визначити унікальні торгові пропозиції та виробити конкурентні стратегії ціноутворення.